checkST2110.js

```diff
@@ -971,210 +1164,481 @@ const test_30_62_4 = (sdp, params) => {
+      }
+    }
+  }
+  if (params.verbose && errors.length == 0) {
+    console.log('Test Passed: ST 2110-22 Section 7.2 Test 1 - All RFC 9134 Section 7.1 required parameters are present');
+  }
+  return errors;
+};
+
+// ST 2110-22 Section 7.3 Test 1 - Check for mandatory bandwidth-field in correct format
+const test_22_73_1 = (sdp, params) => {
+  let streams = sdp.split(/[\r\n]m=/);
+  let sdpLineNumb = 1;
+  let errors = [];
+
+  for (s in streams) {
+    // First element from sdp.split is the session level section. Just move ahead the sdp line count
+    if(s == 0) {
+      let lines = splitLines(streams[s]);
+      sdpLineNumb += lines.length;
+      continue;
+    }
+    let lines = splitLines(streams[s]);
+    let bandwidthPresent = false;
+    for (let x = 0; x < lines.length; x++) {
+      if (lines[x].startsWith('b=')) {
+        let bandwidthMatch = lines[x].match(bandwidthPattern);
+        if (bandwidthMatch == null) {
+          errors.push(new Error(`Line ${sdpLineNumb}: Bandwidth line must be of the form 'b=<bwtype>:<bandwidth>' as per ST 2110-22 Section 7.3.`));
+          sdpLineNumb++;
+          continue;
+        }
+        if (bandwidthMatch[1] != 'AS') {
+          errors.push(new Error(`Line ${sdpLineNumb}: In 'b=<bwtype>:<bandwidth>' bwtype must be 'AS' as per ST 2110-22 Section 7.3.`));
+        }
+        if (Number.isInteger(+bandwidthMatch[2]) == false) {
+          errors.push(new Error(`Line ${sdpLineNumb}: In 'b=<bwtype>:<bandwidth>' bandwidth must be specified as an integer as per ST 2110-22 Section 7.3.`));
+        }
+        bandwidthPresent = true;
+      }
+      sdpLineNumb++;
+    }
+    if (!bandwidthPresent) {
+      errors.push(new Error(`Media Stream ${s}: Required bandwidth-field \'b=<bwtype>:<bandwidth>\' is missing, as per ST 2110-22 Section 7.3.`));
+    }
+  }
+
+  if (params.verbose && errors.length == 0) {
+    console.log('Test Passed: ST 2110-22 Section 7.3 Test 1 - Check for mandatory bandwidth-field');
+  }
+  return errors;
+};
+
+// ST 2110-22 Section 7.4 Test 1 - Check that framerate is specified by one of accepted methods
+const test_22_74_1 = (sdp, params) => {
+  let streams = sdp.split(/[\r\n]m=/);
+  let sdpLineNumb = 1;
+  let errors = [];
+
+  for (s in streams) {
+    let lines = splitLines(streams[s]);
+    // Check for session level framerate attribute
+    if (s == 0) {
+      for (let x = 0; x < lines.length; x++) {
+        if (lines[x].startsWith('a=framerate')) {
+          errors.push(new Error(`Line ${sdpLineNumb}: Framerate must be a media level attribute as per RFC 4566 Section 6`));
+        }
+        sdpLineNumb++;
+      }
+    }
+    // Check each stream for one and only one framerate specification
+    else {
+      let framerateAttributePresent = false;
+      let framerateParameterPresent = false;
+      // First check if exactframerate is specified as a parameter of fmtp
+      let [mtParams, paramErrors] = extractMTParams(sdp, params);
```

Comment:
For sure, we already gratuitously call `extractMTParams` many many times across the whole test suite, but could we avoid calling it multiple times round this `for (s in streams)` loop just by moving this above the loop, could we?

Reply:
Ok with me